compiler/rustc_mir_transform/src/gvn.rs

```diff
@@ -1844,11 +1870,50 @@ impl<'tcx> MutVisitor<'tcx> for StorageRemover<'tcx> {
         match stmt.kind {
             // When removing storage statements, we need to remove both (#107511).
             StatementKind::StorageLive(l) | StatementKind::StorageDead(l)
-                if self.reused_locals.contains(l) =>
+                if self.storage_to_remove.contains(l) =>
             {
                 stmt.make_nop()
             }
             _ => self.super_statement(stmt, loc),
         }
     }
 }
+
+struct StorageChecker<'a, 'tcx> {
+    reused_locals: &'a DenseBitSet<Local>,
+    storage_to_remove: DenseBitSet<Local>,
+    maybe_uninit: ResultsCursor<'a, 'tcx, MaybeUninitializedLocals>,
+}
+
+impl<'a, 'tcx> Visitor<'tcx> for StorageChecker<'a, 'tcx> {
+    fn visit_local(&mut self, local: Local, context: PlaceContext, location: Location) {
+        match context {
+            // These mutating uses do not require the local to be initialized.
+            PlaceContext::MutatingUse(MutatingUseContext::AsmOutput)
+            | PlaceContext::MutatingUse(MutatingUseContext::Call)
+            | PlaceContext::MutatingUse(MutatingUseContext::Store)
+            | PlaceContext::MutatingUse(MutatingUseContext::Yield)
```

Comment:
They do not require the local to be initialized, but they do require it to have storage. Mixing the two notions makes me uneasy.

Reply:
I can probably add a test that triggers this. Is the right solution to use two separate analysis checks (so for these we'll check the `MaybeStorageDead`)?

Reply:
Actually, thinking about this some more: we know that the `reused_locals` have valid storage before their original assignments.

For any replacements (eg target of the `Store`) the `local` cannot be a reused local anyway.

What do you think?